worker/worker.go

```diff
@@ -1585,3 +1499,256 @@ func isErrHostUnreachable(err error) bool {
+		Size:         res.Object.Size,
+		Metadata:     res.Object.Metadata,
+	}, res, nil
+}
+
+func (w *worker) GetObject(ctx context.Context, bucket, path string, opts api.DownloadObjectOptions) (*api.GetObjectResponse, error) {
+	// head object
+	hor, res, err := w.headObject(ctx, bucket, path, false, api.HeadObjectOptions{
+		IgnoreDelim: opts.IgnoreDelim,
+		Range:       opts.Range,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("couldn't fetch object: %w", err)
+	}
+	obj := *res.Object.Object
+
+	// adjust range
+	if opts.Range == nil {
+		opts.Range = &api.DownloadRange{}
+	}
+	opts.Range.Offset = hor.Range.Offset
+	opts.Range.Length = hor.Range.Length
+
+	// fetch gouging params
+	gp, err := w.bus.GougingParams(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't fetch gouging parameters from bus: %w", err)
+	}
+
+	// fetch all contracts
+	contracts, err := w.bus.Contracts(ctx, api.ContractsOpts{})
+	if err != nil {
+		return nil, fmt.Errorf("couldn't fetch contracts from bus: %w", err)
+	}
+
+	// prepare the content
+	var content io.ReadCloser
+	if opts.Range.Length == 0 || obj.TotalSize() == 0 {
+		// if the object has no content or the requested range is 0, return an
+		// empty reader
+		content = io.NopCloser(bytes.NewReader(nil))
+	} else {
+		// otherwise return a pipe reader
+		downloadFn := func(wr io.Writer, offset, length int64) error {
+			ctx = WithGougingChecker(ctx, w.bus, gp)
+			err = w.downloadManager.DownloadObject(ctx, wr, obj, uint64(offset), uint64(length), contracts)
+			if err != nil {
+				w.logger.Error(err)
+				if !errors.Is(err, ErrShuttingDown) &&
+					!errors.Is(err, errDownloadCancelled) &&
+					!errors.Is(err, io.ErrClosedPipe) {
+					w.registerAlert(newDownloadFailedAlert(bucket, path, opts.Prefix, opts.Marker, offset, length, int64(len(contracts)), err))
+				}
+				return fmt.Errorf("failed to download object: %w", err)
+			}
+			return nil
+		}
+		pr, pw := io.Pipe()
+		go func() {
+			err := downloadFn(pw, opts.Range.Offset, opts.Range.Length)
+			pw.CloseWithError(err)
+		}()
+		content = pr
+	}
+
+	return &api.GetObjectResponse{
+		Content:            content,
+		HeadObjectResponse: *hor,
+	}, nil
+}
+
+func (w *worker) HeadObject(ctx context.Context, bucket, path string, opts api.HeadObjectOptions) (*api.HeadObjectResponse, error) {
+	res, _, err := w.headObject(ctx, bucket, path, true, opts)
+	return res, err
+}
+
+func (w *worker) UploadObject(ctx context.Context, r io.Reader, bucket, path string, opts api.UploadObjectOptions) (*api.UploadObjectResponse, error) {
+	// return early if the bucket does not exist
+	_, err := w.bus.Bucket(ctx, bucket)
+	if err != nil {
+		return nil, fmt.Errorf("bucket '%s' not found; %w", bucket, err)
+	}
+
+	// fetch the upload parameters
+	up, err := w.bus.UploadParams(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't fetch upload parameters from bus: %w", err)
+	} else if opts.ContractSet != "" {
```

Comment:
Not really necessary but we could extract the action of applying these two option types to the upload parameters and then validate the upload parameters. It's not awful but there's some code duplication here with the multipart handler. Maybe `func (w *worker) uploadOptions(api.UploadObjectOptions) ([]UploadOption, error)` and a multipart version. Just a thought, feel free to resolve.